### ISSUE TITLE
Why Isn't tf.nn.sampled_softmax_loss used in Transformer Training?

### ISSUE THREAD

**Issue by NickShahML** [tensorflow/tensor2tensor#352]
Thanks for the amazing repo. Really appreciate open-sourcing this work.

I was investigating a few things and it seems that none of the models use ["sampled_softmax"](https://www.tensorflow.org/api_docs/python/tf/nn/sampled_softmax_loss)
and instead, use vanilla softmax_cross_entropy_loss.

Sampled softmax is significantly faster in training and also takes significantly less gpu memory. So I guess my main question is: Why did you guys choose to not use it? It seems to me that you would only stand to gain by using this function. This would benefit even in the scenario of 32k vocab size for your subword datasets.

**Comment by lukaszkaiser** [comment]
Hi Nick. We tried sampled softmax, it wasn't much faster for 32k vocab, and it gives worse results. Using word-pieces (subword tokenizer) removes the need for really large vocab, and we shard the softmax making it less of an issue, so sampled softmax didn't look like sth we need.

**Comment by NickShahML** [comment]
@lukaszkaiser Thanks for letting me know. Makes sense and I appreciate the clarification!
